### PR TITLE
Adelina9 assignment2

### PR DESCRIPTION
**Pull request by Adelina9**

### CODE REVIEW

File: Adelina9/assignment-2/BinaryTree.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Queue;
+import java.util.ArrayList;
+
+public class BinaryTree {
+ 
+    Node root;
+
+    // BFS
+    public Node findNode(int key) {
+        Node n = null;
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: This is basically an alias for null, why not to return it directly instead?

File: Adelina9/assignment-2/BinaryTree.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    public static boolean findAncestors(Node root, Node child, ArrayList<Node> ancestors)
+	{
+		// base case
+		if (root == null) {
+			return false;
+		}
+
+		// given node is found
+		if (root.key == child.key) {
+            // consider the node as a parent to itself
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: Please, fix formatting here.

File: Adelina9/assignment-2/BinaryTree.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     
+            if (node.right!= null) {
+                nodes.add(node.right);
+            }
+        }
+
+        return n;
+    }
+
+    public static boolean findAncestors(Node root, Node child, ArrayList<Node> ancestors)
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: In general, it's much better when function "findSmth" returns this smth. You can use Exceptions to catch incorrect calls (e.g. when node is not present in this tree), that would be more Java-style.

File: Adelina9/assignment-2/BinaryTree.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    
+    public Node findCommonAncestor(BinaryTree tree, Node node1, Node node2) {
+        ArrayList<Node> ancestors1 = new ArrayList<Node>();
+        ArrayList<Node> ancestors2 = new ArrayList<Node>();
+
+        findAncestors(tree.root, node1, ancestors1);
+        findAncestors(tree.root, node2, ancestors2);
+        
+        Node lastNode;
+        // worst case the lowest common ancestor is the root
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: This is not completely correct - you didn't check if the nodes are present in the same tree, and the callAncestors would return empty list then.

File: Adelina9/assignment-2/FindFunctions.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+public Node findNode(int key) {
```
Comment (TDiva) [inline Adelina9/assignment-2/FindFunctions.java]: I believe this shouldn't be here?

File: Adelina9/assignment-2/Main.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import com.sun.javafx.print.PrintHelper.PrintAccessor;
+
+
+public final class Main {
```
Comment (TDiva) [inline Adelina9/assignment-2/Main.java]: We discussed that on our group meeting: there's no real need to write main classes - you can write tests to check if it works, and it would even better guarantee and it will be checked automatically ;)

File: Adelina9/assignment-2/Node.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+class Node {
+    int key;
```
Comment (TDiva) [inline Adelina9/assignment-2/Node.java]: In Java, we prefer to explicitly declare class fields _private_, and generate getters/setters for them. 
https://javarevisited.blogspot.com/2012/03/private-in-java-why-should-you-always.html

File: Adelina9/assignment-2/BinaryTree.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        }
+
+        throw (new NoSuchElementException("Node not in the tree"));
+    }
+
+    public ArrayList<Node> findAncestors(Node root, Node child) throws NoSuchElementException {
+        ArrayList<Node> ancestors = new ArrayList<Node>();
+        //check if the node is in the tree first
+        findNode(child.getKey());
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: There's no real need to do this check in advance, if your private method does it as well.
Also, generally it's better to use functions for their primary purpose, when result value is used.

File: Adelina9/assignment-2/BinaryTree.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+        ancestors1 = findAncestors(tree.root, node1);
+        ancestors2 = findAncestors(tree.root, node2);
+    
+        Node lastNode;
+        Node commonAncestor = tree.root;
+        int found = 0;
+
+        if (ancestors1.size() < ancestors2.size()) {
+            for (int j = 0; j < ancestors1.size(); j++) {
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: Maybe we can find the last common node faster than O(N^2)?
Try to look into examples and you'll find that two loops are not necessary here.

File: Adelina9/assignment-2/Main.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            printAncestors(ancestors1);
+
+            ArrayList<Node> ancestors2 = new ArrayList<Node>();
+            ancestors2 = bt.findAncestors(bt.getRoot(), new Node(key2));
+            printAncestors(ancestors2);
+
+            Node commonAncestor = bt.findCommonAncestor(bt, new Node(key1), new Node(key2));
+            System.out.println("Lowest common ancestor: " + commonAncestor.getKey());
+            
+        } catch (Exception e) {
```
Comment (TDiva) [inline Adelina9/assignment-2/Main.java]: Catching generic Exception is a dangerous move: https://www.javaworld.com/article/2073800/beware-the-dangers-of-generic-exceptions.html

In this case, if you just drop these lines the behavior will be the same - exceptions are propagated upstream until they find the proper "catch" statement.

File: Adelina9/assignment-2/BinaryTree.java (hunk 1 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Got rid of the alias.

File: Adelina9/assignment-2/BinaryTree.java (hunk 2 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Fixed.

File: Adelina9/assignment-2/BinaryTree.java (hunk 3 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Added a function that calls this one and returns the arraylist of ancestors.

File: Adelina9/assignment-2/BinaryTree.java (hunk 4 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: If the nodes are not in the same tree, an exception is thrown.

File: Adelina9/assignment-2/FindFunctions.java (hunk 5 above)
Comment (Adelina9) [reply Adelina9/assignment-2/FindFunctions.java]: Fixed.

File: Adelina9/assignment-2/Node.java (hunk 7 above)
Comment (Adelina9) [reply Adelina9/assignment-2/Node.java]: Added setters and getters and made class fields private.

File: Adelina9/assignment-2/BinaryTree.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    public void setRoot(Node root) {
+        this.root = root;
+    }
+
+    public Node getRoot() {
+        return root;
+    }
+
+    // BFS
+    public Node findNode(int key) throws NoSuchElementException, NullPointerException {
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/Main.java (hunk 10 above)
Comment (Adelina9) [reply Adelina9/assignment-2/Main.java]: Done.

File: Adelina9/assignment-2/BinaryTree.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+    public void setRoot(Node root) {
+        this.root = root;
+    }
+
+    public Node getRoot() {
+        return root;
+    }
+
+    // BFS
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Added javadoc comments.

File: Adelina9/assignment-2/Main.java (hunk 6 above)
Comment (Adelina9) [reply Adelina9/assignment-2/Main.java]: Deleted main class, added automated tests.

File: Adelina9/assignment-2/BinaryTree.java (hunk 9 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Modified the function. Now it has O(n) time complexity.

File: Adelina9/assignment-2/BinaryTree.java (hunk 8 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: The private method does not verify this.

Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/BinaryTree.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    /**
+     * <p>This methods sets the root of the binary tree.</p>
+     * @param root node that will be root
+     */
+    public void setRoot(Node root) {
+        this.root = root;
+    }
+
+    /**
+     * <p>This method uses a BFS traversal to find a node with a
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: Great, this one well explains what the method is doing!

File: Adelina9/assignment-2/BinaryTree.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * <p>This method finds the ancestors of a given node.</p>
+     * @param child node for which we search the ancestors
+     * @return ancestors of the node
+     */
+    public ArrayList<Node> findAncestors(Node child) throws NoSuchElementException {
+        ArrayList<Node> ancestors = new ArrayList<Node>();
+
+        if (findAncestors(this.root, child, ancestors) == false)
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: You can use (!findAncestors...).

File: Adelina9/assignment-2/BinaryTree.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    
+        Node commonAncestor = null;
+        int j = ancestors2.size() - 1;
+
+        for (int i = ancestors1.size() - 1; i >= 0; i--) {
+            if (ancestors1.get(i).getKey() == ancestors2.get(j).getKey()) {
+                commonAncestor = ancestors1.get(i);
+                // the second arraylist can be shorter 
+                // verification is required to avoid out of bounds exception
+                if (j - 1 >= 0)
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: this check can be either `j>=1` or even 
``` 
j--
if (j == 0) break;
```

File: Adelina9/assignment-2/BinaryTree.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * <p>This method finds the lowest common ancestor of two given nodes.</p>
+     * @param node1 first node
+     * @param node2 second node
+     * @return lowest common ancestor of the two nodes
+     */
+    public Node findCommonAncestor(Node node1, Node node2) throws NoSuchElementException {
+        ArrayList<Node> ancestors1 = new ArrayList<Node>();
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: Why do you need to initialize this list?
Another thing, it's a good practice to use Interfaces when you specify variables:
```
List<String> someList = new ArrayList<>();
```

File: Adelina9/assignment-2/BinaryTree.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+        throw (new NoSuchElementException("Node not in the tree"));
+    }
+
+    /**
+     * <p>This method finds the ancestors of a given node.</p>
+     * @param child node for which we search the ancestors
+     * @return ancestors of the node
+     */
+    public ArrayList<Node> findAncestors(Node child) throws NoSuchElementException {
```
Comment (TDiva) [inline Adelina9/assignment-2/BinaryTree.java]: Prefer using interfaces when you specify variables/return types. You don't need to guarantee that this method will return ArrayList, it can be any List.

File: Adelina9/assignment-2/Node.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
+public class Node {
```
Comment (TDiva) [inline Adelina9/assignment-2/Node.java]: \+ Javadoc for the class.

File: Adelina9/assignment-2/README.md [hunk 19]
```diff
@@ -1,12 +1,12 @@
-# Assignment 2
```
Comment (TDiva) [inline Adelina9/assignment-2/README.md]: This is a weird change, try to keep track of what gets committed and avoid unnecessary changes.

File: Adelina9/assignment-2/TestBinaryTree.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+   @Test(expected = NoSuchElementException.class)
+   public void testFindUnexistingNode() {
+      bt.findNode(12);
+      bt.findNode(132);
+   }
+
+   @Test
+   public void testFindAncestors() {
+      ArrayList<Node> ancestors = new ArrayList<Node>();
```
Comment (TDiva) [inline Adelina9/assignment-2/TestBinaryTree.java]: The same as before, no need to initialize this list with empty.

File: Adelina9/assignment-2/TestBinaryTree.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+      ancestors = bt.findAncestors(new Node(2));
+
+      ArrayList<Node> expectedAncestors = new ArrayList<Node>();
```
Comment (TDiva) [inline Adelina9/assignment-2/TestBinaryTree.java]: You can use Arrays.asList(...) method: https://www.geeksforgeeks.org/arrays-aslist-method-in-java-with-examples/
This is the benefit of using generics as variable type.

File: Adelina9/assignment-2/TestBinaryTree.java [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+      expectedAncestors.add(bt.findNode(2));
+      expectedAncestors.add(bt.findNode(3));
+      expectedAncestors.add(bt.findNode(7));
+
+      for (int i = 0; i < ancestors.size(); i++) {
+         assertEquals(expectedAncestors.get(i).getKey(), ancestors.get(i).getKey());
```
Comment (TDiva) [inline Adelina9/assignment-2/TestBinaryTree.java]: You can compare lists without comparing each element individually - list has equals method which checks elements and order.
```
assertEquals(expectedAncestors, ancestors);
```

To make sure you compare keys only you can override `equals()` method in Node class. Or you can prepare a list of nodes from the test tree itself, to be even more precise.

File: Adelina9/assignment-2/BinaryTree.java [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.ArrayList;
+
+public class BinaryTree {
+ 
+    private Node root;
+
+    /**
+     * <p>BinaryTree constructor</p>
```
[the remark this answers is not in the document]
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/BinaryTree.java (hunk 14 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/BinaryTree.java (hunk 15 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/BinaryTree.java (hunk 16 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Use interfaces now, also removed the useless initialisation.

File: Adelina9/assignment-2/BinaryTree.java (hunk 17 above)
Comment (Adelina9) [reply Adelina9/assignment-2/BinaryTree.java]: Done.

File: Adelina9/assignment-2/Node.java (hunk 18 above)
Comment (Adelina9) [reply Adelina9/assignment-2/Node.java]: Done.

File: Adelina9/assignment-2/README.md (hunk 19 above)
Comment (Adelina9) [reply Adelina9/assignment-2/README.md]: Will do.

File: Adelina9/assignment-2/TestBinaryTree.java (hunk 20 above)
Comment (Adelina9) [reply Adelina9/assignment-2/TestBinaryTree.java]: Done.

File: Adelina9/assignment-2/TestBinaryTree.java (hunk 21 above)
Comment (Adelina9) [reply Adelina9/assignment-2/TestBinaryTree.java]: Used Arrays.asList to create the list of nodes.

File: Adelina9/assignment-2/TestBinaryTree.java (hunk 22 above)
Comment (Adelina9) [reply Adelina9/assignment-2/TestBinaryTree.java]: Overrided equals to compare only keys.
